MediaBrowser.Common/Extensions/ProcessExtensions.cs

```diff
@@ -0,0 +1,79 @@
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace MediaBrowser.Common.Extensions
+{
+    /// <summary>
+    /// Extension methods for <see cref="Process"/>.
+    /// </summary>
+    public static class ProcessExtensions
+    {
+        /// <summary>
+        /// Asynchronously wait for the process to exit.
+        /// </summary>
+        /// <param name="process">The process to wait for.</param>
+        /// <param name="timeout">The duration to wait before cancelling waiting for the task.</param>
+        /// <returns>True if the task exited normally, false if the timeout elapsed before the process exited.</returns>
+        public static async Task<bool> WaitForExitAsync(this Process process, TimeSpan timeout)
+        {
+            using (var cancelTokenSource = new CancellationTokenSource(timeout))
+            {
+                return await WaitForExitAsync(process, cancelTokenSource.Token);
+            }
+        }
+
+        /// <summary>
+        /// Asynchronously wait for the process to exit.
+        /// </summary>
+        /// <param name="process">The process to wait for.</param>
+        /// <param name="cancelToken">A <see cref="CancellationToken"/> to observe while waiting for the process to exit.</param>
+        /// <returns>True if the task exited normally, false if cancelled before the process exited.</returns>
+        public static async Task<bool> WaitForExitAsync(this Process process, CancellationToken cancelToken)
+        {
+            if (!process.EnableRaisingEvents)
+            {
+                throw new InvalidOperationException("EnableRisingEvents must be enabled to async wait for a task to exit.");
+            }
+
+            // Add an event handler for the process exit event
+            var tcs = new TaskCompletionSource<bool>();
+            process.Exited += (sender, args) => tcs.TrySetResult(true);
+
+            // Return immediately if the process has already exited
+            if (process.HasExitedSafe())
+            {
+                return true;
+            }
```

Comment:
move this above allocation of the completion source - why add more garbage for GC? :)

Reply:
I think this has to go after the event is attached. Otherwise there is a race condition where the process could exit between calling `process.HasExitedSafe()` and `process.Exited += ...` and you'll end up in a deadlock. I might be mistaken though, concurrency is hard 😅

Reply:
I don't think I understand what you mean... My proposal is to do it like
```cs
            // Return immediately if the process has already exited
            if (process.HasExitedSafe())
            {
                return true;
            }

            // Add an event handler for the process exit event
            var tcs = new TaskCompletionSource<bool>();
            process.Exited += (sender, args) => tcs.TrySetResult(true);
```
where is race condition?..

Reply:
```cs
            // Return immediately if the process has already exited
            if (process.HasExitedSafe())
            {
                return true;
            }

            // Current thread could be preempted here and then the spawned process exits before it resumes again
            // In this case the Exited event handler hasn't been attached yet, and will never be raised when we attach it below

            // Add an event handler for the process exit event
            var tcs = new TaskCompletionSource<bool>();
            process.Exited += (sender, args) => tcs.TrySetResult(true);
```

Reply:
I see. Thanks for explanation!